readme.md
# Knock

PHP authorization class for logging in and logging out. No form is included.

*Version 2.1* [Changelog](changelog.md)

## In short

- Persistent cookie
- Whitelist of IPs
- Plenty of options
- No dependencies
- No database

## Usage

### Create a user

The folder structure may look like below where filename should be `[username].php`.

```text
└─ users
   └─ test@example.com.php
```

#### Inside the user file

- Inside the user file you return an array including the password.
- To use the password `test` you need to hash it with a tool like [SHA256 online hash function](https://emn178.github.io/online-tools/sha256.html).
- *You can also use `hash('sha256', 'test')` as the password, but that is not recommended. Use it for testing purposes only!*

```php
<?php return [
  'password' => '9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08',
];
```

## Methods

All methods work simliar to below.

```php
include __DIR__ . '/knock.php';

$_POST['username'] = 'test@example.com';
$_POST['password'] = 'test';

$knock = new Knock();
if($knock->login()) {
  echo 'You are logged in';
} else {
  print_r($knock->results); // ['success' => false, 'error' => 'error_message']
}
```

| Name                  | Args     | Default                                   | Description                                                                                |
| --------------------- | -------- | ----------------------------------------- | ------------------------------------------------------------------------------------------ |
| `isAuthorized()`      | -        | -                                         | Returns `true` if the user `$_POST` password matches the user file password                |
| `isLoggedIn()`        | -        | -                                         | Returns `true` if the user `$_COOKIE` hash matches the temp file hash (which is salted)    |
| `getCookieExpires()`  | -        | -                                         | Returns the cookie expires timestamp if it exists                                          |
| `keepAlive()`         | -        | -                                         | Will run `refresh`, but only if the cookie is close to its expire timestamp                |
| `login()`             | -        | -                                         | Login a user from `$_POST['username']` and `$_POST['password']` if you use these keys      |
| `logout()`            | -        | -                                         | It will delete the `$_COOKIE` and remove the temp file                                     |
| `refresh()`           | -        | -                                         | If logged in, it will create a new hash and expire timestamp                               |

## Options (optional)

To use the options you need to place a `options.php` file in the root.

**Defaults**

```php
return [
  'algorithm' => 'sha256',
  'cookie_prefix' => 'knock',
  'cookie_refresh' => 15,
  'login_delay' => 500,
  'key_cookie_expires' => 'expires',
  'key_cookie_hash' => 'hash',
  'key_cookie_username' => 'username',
  'key_post_password' => 'password',
  'key_post_username' => 'username',
  'path_temp' => __DIR__ . '/temp/',
  'path_users' => __DIR__ . '/users/',
  'salt' => '',
  'setcookie_domain' => '',
  'setcookie_expires' => 0,
  'setcookie_httponly' => false,
  'setcookie_path' => '',
  'setcookie_secure' => false,
  'whitelist' => [],
];
```

### Explained

<!--| `login_attempts`      | integer  | `5`                                       | Not yet implemented                                                                        |-->

| Name                  | Type     | Default                                   | Description                                                                                |
| --------------------- | -------- | ----------------------------------------- | ------------------------------------------------------------------------------------------ |
| `algorithm`           | string   | `'sha256'`                                | The algorithm used to create hashes                                                        |
| `cookie_prefix`       | string   | `'knock'`                                 | To prevent collisions with other cookies you can set your own prefix.                      |
| `cookie_refresh`      | string   | `15`                                      | When using `knock::keepAlive()` this value is used to decide when to refresh the cookie.   |
| `login_delay`         | integer  | `500`                                     | A millisecond number to delay the authorization. It will prevent bruce force attacks       |
| `key_cookie_expires`  | string   | `'expires'`                               | Change this to make the cookie a bit more cryptic.                                         |
| `key_cookie_hash`     | string   | `'hash'`                                  | Change this to make the cookie a bit more cryptic.                                         |
| `key_cookie_username` | string   | `'username'`                              | Change this to make the cookie a bit more cryptic.                                         |
| `key_post_password`   | string   | `'password'`                              | Change this to make the post a bit more cryptic.                                           |
| `key_post_username`   | string   | `'username'`                              | Change this to make the post a bit more cryptic.                                           |
| `path_temp`           | string   | `__DIR__ . '/users/'`                     | Path where temporary login data is stored                                                  |
| `path_users`          | string   | `__DIR__ . '/temp/'`                      | Path where user files are stored                                                           |
| `salt`                | string   | `''`                                      | A random string that will be added to the temp file. It will make it a bit harder to hack  |
| `setcookie_domain`    | string   | `''`                                      | See [setcookie](http://php.net/manual/en/function.setcookie.php)                           |
| `setcookie_expires`   | integer  | `0`                                       | See [setcookie](http://php.net/manual/en/function.setcookie.php)                           |
| `setcookie_httponly`  | boolean  | `false`                                   | See [setcookie](http://php.net/manual/en/function.setcookie.php)                           |
| `setcookie_path`      | string   | `''`                                      | See [setcookie](http://php.net/manual/en/function.setcookie.php)                           |
| `setcookie_secure`    | string   | `false`                                   | See [setcookie](http://php.net/manual/en/function.setcookie.php)                           |
| `whitelist`           | array    | `[]`                                      | Allwed IP numbers. If not set, all are allowed. Ending wildcard `*` supported.             |                                                        |

## Security headers

Make sure to use security headers on pages where you call Knock.

```php
header("X-Frame-Options: sameorigin"); // Prevent iframe access
header("X-XSS-Protection: 1; mode=block"); // XSS protection
header("X-Content-Type-Options: nosniff"); // Require correct MIME type for CSS and JS
header("Content-Security-Policy: default-src 'none'; script-src 'self'; connect-src 'self'; img-src 'self'; style-src 'self';");
header("Referrer-Policy: no-referrer");
```

**Source:** https://zinoui.com/blog/security-http-headers

## Generate strong passwords

The probably best service out there to generate passwords is https://www.expressvpn.com/password-generator.

## Hacker challenge

Do you think you can hack this thing? I would appreciate if you tried. If you succeed, report in an issue what you did.

Your report should contain a real life case, not a theoretically one. An example of a theoretically hack would be to guess the cookie username and hash. Because even the cookie keys are unknown and different for each installation, it would take the sun to go out before your guess is correct.

I will not pay you anything for the work, but you can get a mention in the readme file and perhaps a link to your site as thanks (if it's not unhealthy).

## Requirements

- PHP 7+

## Disclaimer

This plugin is provided "as is" with no guarantee. Use it at your own risk and always test it yourself before using it in a production environment. If you find any issues, please [create a new issue](issues/new).

## Donate

Donate to [DevoneraAB](https://www.paypal.me/DevoneraAB) if you want.

## License

MIT

changelog.md
# Changelog

## 2.1

- Fixed bug with keepAlive

## 2.0

- Updated TinyOptions
- Added `getOptions`, `getUsername` and `getUserpath` methods

## 1.9

- Almost complete rewrite into multiple classes
- New tests
- Options is no longer called from a file, but from an input array

## 1.8

- Bug fixes
- Improved error handling

## 1.7

- Added "Hacker challange" to readme.
- Changed how user files work. Now it's an array instead of a string. Will be useful in the future to be able to add roles etc.
- Added better error handling.
- Changed syntax and skipped static methods.
- Removed callbacks. No need for them as methods return true or false.

## 1.6

- Rewritten codebase.
- All option keys has changed. See docs.
- Most default option values has been changed. See docs.
- Added `algorithm`, `whitelist` and `setcookie_httponly` options.

## 1.5

- Added expires cookie to be able to see when the cookie expires.
- Added cookie option `cookie.expires.key` to make the expires cookie a bit more cryptic.
- Added cookie option `cookie.refresh` in case `knock::keepAlive()` is used.
- Added function `knock::keepAlive()` to trigger the cookies to refreh itself, if the cookie time is close to expire.
- Added function `knock::refresh()` to trigger the cookies to refresh.
- Added function `knock::getCookieExpires()` to get the cookies expire timestamp.
- Updated all tests.

## 1.4

- Added post option `post.username.key` and `post.password.key` to make it less guess fiendly.
- Changed the default cookie expire time from 20 years to 2 days.

## 1.3

- Added `salt` option.

## 1.2

- Force cookie to `httponly` is now `true`.
- Added cookie options `cookie.domain` and `cookie.secure`.
- Added `cookie.username.key` and `cookie.hash.key` options to have a bit more cryptic cookies.

## 1.1

- If temp folder is missing it's now added.
- Changed option `prefix` to `cookie.prefix`
- Check if user file exists before trying to use the file
- Option `delay` implemented. It will delay the login time to prevent brute force attacks

## 1.0

- Initial release

knock.php
<?php
class Knock {
  private $error = null;

  // Set options
  public function __construct($options = null) {
    $this->options($options);
  }

  ## PRIVATE

  // Defaults
  private function defaults() {
    return [
      'algorithm' => 'sha256',
      'cookie_prefix' => 'knock',
      'cookie_refresh' => 15, // Bad key name
      'login_delay' => 500,
      'login_attempts' => 5,
      'key_cookie_expires' => 'expires',
      'key_cookie_hash' => 'hash',
      'key_cookie_username' => 'username',
      'key_post_password' => 'password',
      'key_post_username' => 'username',
      'path_temp' => __DIR__ . '/temp/',
      'path_users' => __DIR__ . '/users/',
      'salt' => '',
      'setcookie_domain' => '',
      'setcookie_expires' => 0,
      'setcookie_httponly' => false,
      'setcookie_path' => '',
      'setcookie_secure' => false,
      'whitelist' => [],
    ];
  }

  // Set options
  private function options($options) {
    $array = $this->defaults();

    if($options) {
      $array = array_merge($this->defaults(), $options);
    }

    foreach($array as $key => $value) {
      $this->{$key} = $value;
    }    
  }

  // Login user
  private function loginUser($username) {
    $hash = bin2hex(random_bytes(16));

    if(!$this->writeCookie($hash, $username)) return;
    if(!$this->writeFile($hash, $username)) return;
    return true;
  }

  // Logout user
  private function logoutUser($username) {
    $this->deleteCookies();

    if($username) {
      $path = $this->path_temp . $username . '.php';
      return $this->deleteFile($path);
    }
    return true;
  }

  // Delete file
  private function deleteFile($path) {
    echo $path;
    if(file_exists($path)) {
      if(!unlink($path)) {
        $this->error = [1, $path];
        return;
      }
    }
    return true;
  }

  // Make directiory
  private function makeDir($path) {
    if(!file_exists($path)) {
      if(!mkdir($path)) {
        $this->error = [2, $path];
        return;
      }
    }
    return true;
  }

  // Delete cookies
  private function deleteCookies() {
    if(!$this->setCookie('[' . $this->key_cookie_username . ']', '', 0)) return;
    if(!$this->setCookie('[' . $this->key_cookie_hash . ']', '', 0)) return;
    if(!$this->setCookie('[' . $this->key_cookie_expires . ']', '', 0)) return;

    return true;
  }

  // Write cookie on login
  private function writeCookie($hash, $username) {
    if(!$this->setCookie('[' . $this->key_cookie_username . ']', $username)) return;
    if(!$this->setCookie('[' . $this->key_cookie_hash . ']', $hash)) return;
    if(!$this->setCookie('[' . $this->key_cookie_expires . ']', $this->setcookie_expires)) return;

    return true;
  }

  // Set cookie
  private function setCookie($key, $value, $expires = null) {
    $expires = ($expires === null) ? $this->setcookie_expires : $expires;

    $success = setcookie(
      $this->cookie_prefix . $key,
      $value,
      $expires,
      $this->setcookie_path,
      $this->setcookie_domain,
      $this->setcookie_secure,
      $this->setcookie_httponly
    );

    if(!$success) {
      $this->error = [3, $key];
    }
    return $success;
  }

  // Write temp file to disc
  private function writeFile($hash, $username) {
    if(!$this->makeDir($this->path_temp)) return;

    $content = "<?php return '" . $this->fileHash($hash) . "';";
    $success = file_put_contents($this->path_temp . $username . '.php', $content);

    if(!$success) {
      $this->error = [4, $path];
    }
    return $success;
  }

  private function fileHash($hash) {
    return hash($this->algorithm, $hash . $this->salt);
  }

  // Check if IP is allowed
  private function ipAllowed() {
    if(!empty($this->whitelist)) {
      foreach($this->whitelist as $item) {
        if(substr($item, -1) === '*') {
          $starts_with = substr($item, 0, -1);
          if(strpos($_SERVER['REMOTE_ADDR'], $starts_with) === 0) {
            return true;
          }
        } elseif($item === $_SERVER['REMOTE_ADDR']) {
          return true;
        }
      }
      $this->error = [5];
      return;
    }
  }

  // Filepath
  private function filepath($type, $username = null) {
    switch($type) {
      case 'post':
        $filepath = $this->path_users . $_POST[$this->key_post_username] . '.php';
        break;
      case 'cookie':
        $filepath = $this->path_temp . $_COOKIE[$this->cookie_prefix][$this->key_cookie_username] . '.php';
        break;
      case 'user':
        $filepath = $this->path_temp . $username . '.php';
        break;
    }

    if(!file_exists($filepath)) {
      $this->error = [6, $filepath];
      return;
    } else {
      return $filepath;
    }
  }

  // Get cookie
  private function getCookie($key) {
    $key = 'key_' . $key;
    if(isset($_COOKIE[$this->cookie_prefix][$this->{$key}])) {
      return $_COOKIE[$this->cookie_prefix][$this->{$key}];
    }
  }

  // Has post
  private function hasPost($key) {
    $key = 'key_' . $key;
    if(!isset($_POST[$this->{$key}])) {
      $this->error = [8, $key];
      return;
    }
    return true;
  }

  private function hasCookies() {
    if(!$this->getCookie('cookie_expires')) return;
    if(!$this->getCookie('cookie_hash')) return;
    if(!$this->getCookie('cookie_username')) return;

    return true;
  }

  // Check if user is authorized with post variables
  private function userIsAuthorized() {
    if(!$this->hasPost('post_username')) return;
    if(!$this->hasPost('post_password')) return;
    if(!$this->filepath('post')) return;

    $data = include($this->filepath('post'));

    if(!isset($data['password'])) {
      $this->error = [9];
      return;
    }
    $post_password = hash($this->algorithm, $_POST[$this->key_post_password]);

    if($data['password'] !== $post_password) {
      $this->error = [10];
      return;
    }
    return true;
  }

  // Results
  private function setResults() {
    $results['success'] = ($this->error === null) ? true : false;

    if($this->error !== null) {
      $results['error'] = $this->error;
    }
    $this->results = $results;
  }  

  ## PUBLIC

  // is authorized
  public function isAuthorized() {
    $this->userIsAuthorized();
    $this->setResults();
    return $this->results['success'];
  }

  // Check if user is logged in with cookie
  public function isLoggedIn() {
    if(!$this->hasCookies()) return;
    if(!$this->filepath('cookie')) return;

    $hash = include($this->filepath('cookie'));
    $hash_cookie = $this->getCookie('cookie_hash');
    $hash_cookie = hash($this->algorithm, $hash_cookie . $this->salt);

    if($hash != $hash_cookie) {
      $this->error = [11];
    }

    $this->setResults();
    return $this->results['success'];
  }

  // Get the cookies expire timestamp
  public function getCookieExpires() {
    return $this->getCookie('cookie_expires');
  }

  // Create user
  public function createUser($username = null, $password = null) {
    if(!$username || !$password) {
      $this->error = [12];
    } elseif(file_exists($this->path_users . $username . '.php')) {
      $this->error = [13];
    } else {
      $content = sprintf("<?php return ['password' => '%s'];", hash('sha256', $password));
      $success = file_put_contents($this->path_users . $username . '.php', $content);
      if(!$success) {
        $this->error = [14];
      }
    }
    $this->setResults();
    return $this->results['success'];
  }

  // Refresh the cookies if the cookies will soon expire
  public function keepAlive() {
    $minutes = round(((int)$this->getCookieExpires()-time())/60);
    $diff = $minutes - $this->cookie_refresh;
    return ($diff < 0) ? $this->refresh() : true;
  }

  // Login with post variables
  public function login() {
    usleep($this->login_delay * 1000);
    
    if($this->ipAllowed() && $this->userIsAuthorized()) {
      $this->loginUser($_POST[$this->key_post_username]);
    }
    $this->setResults();
    return $this->results['success'];
  }

  // Logout
  public function logout() {
    $this->logoutUser($this->getCookie('cookie_username'));
    $this->setResults();
    print_r($this->results);
    return $this->results['success'];
  }

  // Refresh the cookies, creates new hash and expire timestamp
  public function refresh() {
    if($this->isLoggedIn()) {
      return $this->loginUser($this->getCookie('cookie_username'));
    }
    return true;
  }
}